### DISCUSSION TITLE
How can I set an overall brightness?

### DISCUSSION THREAD

**Discussion by ardumiguel** [Makuna/NeoPixelBus#761] [Q&A] [answered]
Just comming from normal Adafruit NeoPixel library, and so im used to working with original neopixel's `setbrightness()`.

I tried various things but none of them seem to work with my code, which involves `GetPixelColor(0);` in order to fade between colors.

Any thoughts?

**Comment by Makuna** [comment, marked as answer]
https://github.com/Makuna/NeoPixelBus/wiki/FAQ-%239
There is an example using NeoPixelBusLg and it has replaced NeoPixelBrightnessBus so that FAQ needs to get updated.
Key thing is, GetPixelColor() and overall brightness changes are fraught with hidden problems that too many users kept hitting and not understanding.  

So, if you need GetPixelColor to work (provide the original color without the brightness applied) then the recommended way is demonstrated in the NeoPixelDibTest example.

Create a buffer, treat it as your strip for your animations and calls to GetPixelColor/SetPixelColor.  Then Render that buffer into the NeoPixelBus object.

**Reply by ardumiguel** [reply]
Alright, thank you. Then it was just a misunderstanding from my side. I didn't know `GetPixelColour()` was gonna return untouched values.

**Reply by Makuna** [reply]
If you use NeoPixelBusLg, the GetPixelColor will return the post dim value.
If you use NeoPixelBrightnessBus, the GetPixelColor will return a value that is trying to approximate the original color that was set using SetPixelColor.  BUT, the lower the brightness, the less accurate the returned color will match the original.  This is exactly what Adafruits library did, but the nuance of not being accurate was never mentioned and caused issues for users.

**Reply by ardumiguel** [reply]
Then, is it normal that after trying neopixelbuslg and retrieving pixel info I still get a jump when fading from current colour to desired one? As if getpixelcolour wasn't returning post dim colour.

**Reply by Makuna** [reply]
[Wiki on the NeoPixelBusLg](https://github.com/Makuna/NeoPixelBus/wiki/NeoPixelBusLg-object)

That would not be normal to call SetPixelColor and then call GetPixelColor and get back the same color for NeoPixelBusLg unless the luminance was at 255.  Further, if you don't provide a gamma method, it will use a default one even further modifying the value.  [Another Wiki link about gamma.](https://github.com/Makuna/NeoPixelBus/wiki/T_GAMMA)

ALso note, changing the luminance does not affect the colors that were previously set using SetPixelColor, it will only effect those calling SetPixelColor after.  Again, since the NeoPixelBusLg doesn't retain the original color that you called SetPixelColor with, there is no way for it to recalculate with that in mind, all it has is the post values.  So the model is, change luminance, draw all pixels.
